assigner/commands/score.py

```diff
@@ -0,0 +1,418 @@
+    """
+    A utility function for printing an ASCII histogram
+    for a one-dimensional data set
+    """
+    print("ASCII Histogram:")
+    num_buckets = 10
+    range_min = min(scores)
+    range_max = max(scores)
+    max_col = os.get_terminal_size()[0] - 15
+    bucket_width = (range_max - range_min) / num_buckets
+    buckets = [(i * bucket_width, (i + 1) * bucket_width) for i in range(num_buckets)]
+    counts = {}
+
+    # First count up each bucket
+    for bucket in buckets:
+        count = len([score for score in scores if bucket[0] <= score < bucket[1]])
+        # If it's the last bucket we include the top (i.e. the range max)
+        if bucket == buckets[-1]:
+            count += scores.count(range_max)
+        counts[bucket] = count
+
+    # Then set up the scale factor to maximally utilize the terminal space
+    mult_factor = max_col / max(counts.values())
+
+    # Finally, print everything out
+    for bucket in buckets:
+        proportional_len = int(counts[bucket] * mult_factor)
+        print(
+            "[{:4}, {:5}{}: {}".format(
+                bucket[0],
+                bucket[1],
+                (")", "]")[bucket == buckets[-1]],
+                proportional_len * "=",
+            )
+        )
+
+
+def handle_scoring(
+    conf: Config,
+    backend: BackendBase,
+    args: argparse.Namespace,
+    student: Dict[str, Any],
+) -> Optional[float]:
+    """
+    Obtains the autograded score from a repository's CI jobs
+    :param student: The part of the config structure with info
+    on a student's username, ID, and section
+    :return: The score obtained from the results file
+    """
+    hw_name = args.name
+    upload = args.upload if "upload" in args else True
+    files_to_check = set(args.files)
+    backend_conf = conf.backend
+    username = student["username"]
+    student_section = student["section"]
+    full_name = backend.student_repo.build_name(
+        conf.semester, student_section, hw_name, username
+    )
+    try:
+        repo = backend.student_repo(backend_conf, conf.namespace, full_name)
+        logger.info("Scoring %s...", repo.name_with_namespace)
+        if "id" not in student:
+            student["id"] = backend.repo.get_user_id(username, backend_conf)
+        if not args.noverify:
+            unlock_time = repo.get_member_add_date(student["id"])
+            check_repo_integrity(repo, files_to_check, unlock_time)
+        score = get_most_recent_score(repo, args.path)
+        if upload:
+            canvas = OptionalCanvas.get_api(conf)
+            section_ids = OptionalCanvas.get_section_ids(conf, hw_name)
+            assignment_ids = OptionalCanvas.get_assigment_ids(conf, hw_name)
+            course_id = section_ids[student_section]
+            assignment_id = assignment_ids[student_section]
+            try:
+                canvas.put_assignment_submission(
+                    course_id, assignment_id, student["canvas-id"], str(score) + "%"
```

Comment:
What happens if `canvas-id` is not set for a student? (for example, if the student was added to the roster using a prior version of Assigner.)

I could see an argument for fetching canvas ids for students when upgrading configs, but I think that's probably over-complicated and it's better to just assume that some students may not have canvas ids and to warn on that with instructions on how to fetch their id.

Reply:
If a canvas-id doesn't exist for a student, from my understanding the only options are to:
1. Have the user manually look it up and add it to the roster entry in `_config.yml`
2. Have the user manually remove the student from the roster in `_config.yml` and then re-run `assigner canvas import`

The second option is easier than the first but neither of them are ideal.  Did you have any thoughts on a cleaner way of handling this?

Reply:
I think `assigner canvas import --force` should import missing canvas-ids for students in the roster, if I am thinking about it right. Can you double-check that that works?

Reply:
I think that will add duplicate entries to the roster instead of replacing existing ones

Reply:
Dammit you're right.

Reply:
Well, anyhow, at a minimum we can point them to `assigner roster remove` to delete users; given that I anticipate this will not be a frequent occurrence, I'm happy leaving it at that rather than adding more features.